src/refiners/foundationals/latent_diffusion/ella_adapter.py

```diff
@@ -0,0 +1,285 @@
+    ) -> None:
+        super().__init__(
+            fl.UseContext("diffusion", "timestep"),
+            RangeEncoder(time_channel, time_embedding_dim, device=device, dtype=dtype),
+            fl.SetContext("ella", "timestep_embedding"),
+        )
+
+
+class SquaredReLU(fl.ReLU):
+    def __init__(self) -> None:
+        super().__init__()
+
+    def forward(self, x: Tensor) -> Tensor:
+        return super().forward(x).pow(2)
+
+
+class AdaLayerNorm(fl.Chain):
+    def __init__(
+        self,
+        embedding_dim: int,
+        time_embedding_dim: int,
+        device: Device | str | None = None,
+        dtype: DType | None = None,
+    ) -> None:
+        super().__init__(
+            fl.Parallel(
+                LayerNormNoAffine(embedding_dim, eps=1e-6, device=device, dtype=dtype),
+                fl.Chain(
+                    fl.UseContext("ella", "timestep_embedding"),
+                    fl.SiLU(),
+                    fl.Linear(time_embedding_dim, embedding_dim * 2, device=device, dtype=dtype),
+                ),
+            ),
+            fl.Lambda(self._scale_shift_tensors),
+        )
+
+        self._init_parameters()
+
+    def _init_parameters(self) -> None:
+        _linear: fl.Linear = self.ensure_find(fl.Linear)
+        nn.init.zeros_(_linear.weight)
+        nn.init.zeros_(_linear.bias)
+
+    def _scale_shift_tensors(self, x: Tensor, time_embedding: Tensor) -> Tensor:
+        shift, scale = time_embedding.chunk(2, dim=-1)
+        return x * (1 + scale) + shift
+
+
+class ParameterInitialized(fl.Parameter):
+    def __init__(
+        self, *dims: int, requires_grad: bool = True, device: Device | str | None = None, dtype: DType | None = None
+    ) -> None:
+        super().__init__(*dims, requires_grad=requires_grad, device=device, dtype=dtype)
+        nn.init.normal_(self.weight, mean=0, std=dims[1] ** 0.5)
+
+
+class Latents(fl.Chain):
+    def __init__(
+        self,
+        num_latents: int,
+        width: int,
+        device: Device | str | None = None,
+        dtype: DType | None = None,
+    ) -> None:
+        super().__init__(
+            ParameterInitialized(
+                num_latents,
+                width,
+                device=device,
+                dtype=dtype,
+            ),
+        )
+
+
+class PerceiverAttention(fl.Chain):
+    def __init__(
+        self,
+        width: int,
+        num_heads: int,
+        timestep_embedding_dim: int,
+        device: Device | str | None = None,
+        dtype: DType | None = None,
+    ) -> None:
+        super().__init__(
+            fl.Distribute(
+                AdaLayerNorm(width, timestep_embedding_dim, device=device, dtype=dtype),
+                AdaLayerNorm(width, timestep_embedding_dim, device=device, dtype=dtype),
+            ),
+            fl.Parallel(
+                fl.GetArg(index=1),
+                fl.Lambda(func=self.to_kv),
+                fl.Lambda(func=self.to_kv),
+            ),
```

Comment:
nit: slightly hard to read at a glance, could it be simplified ?

Reply:
following Cedric comments, this is almost and identical implementation to IP adapter's PerceiverAttention class. Do you have a suggestion it simplify it ?

Reply:
I guess this version is the less verbose one then